e2e/main.sh

```diff
@@ -0,0 +1,151 @@
+#!/usr/bin/env bash
+
+set -o errtrace
+set -o nounset
+set -o pipefail
+
+SCRIPT_DIR="$(readlink -f "$(dirname "${BASH_SOURCE[0]}")")"
+OUTPUT_DIR="${ARTIFACT_DIR:-_output}/microshift-e2e-$(date +'%Y%m%d-%H%M%S')/"
+[ ! -d "${OUTPUT_DIR}" ] && mkdir -p "${OUTPUT_DIR}"
+
+log() {
+    echo -e "$(date +'%H:%M:%S.%N')    $*"
+}
+
+var_should_not_be_empty() {
+    local var_name=${1}
+    if [ -z "${!var_name+x}" ]; then
+        echo >&2 "Environmental variable '${var_name}' is unset"
+        return 1
+    elif [ -z "${!var_name}" ]; then
+        echo >&2 "Environmental variable '${var_name}' is empty"
+        return 1
+    fi
+}
+
+function_should_be_exported() {
+    local fname=${1}
+    if ! declare -F "$fname"; then
+        echo >&2 "Function '$fname' is unexported. It is expected that function is provided for interacting with cloud provider"
+        return 1
+    fi
+}
+
+prechecks() {
+    var_should_not_be_empty USHIFT_IP || exit 1
+    var_should_not_be_empty USHIFT_USER || exit 1
+    # TODO Check passwordless SSH
+    # TODO Check passwordless sudo
+
+    # Just warning for now
+    # Following functions needed only for runs in CI
+    function_should_be_exported firewall::open_port
+    function_should_be_exported firewall::close_port
+}
+
+microshift_get_konfig() {
+    tmpfile=$(mktemp /tmp/microshift-e2e-konfig.XXXXXX)
+    ssh "$USHIFT_USER@$USHIFT_IP" 'sudo cat /var/lib/microshift/resources/kubeadmin/'"$USHIFT_IP"'/kubeconfig' >"$tmpfile"
+    echo "$tmpfile"
+}
+
+microshift_check_readiness() {
+    local test_output="${1}"
+    log "Waiting for MicroShift to become ready"
+    ssh "$USHIFT_USER@$USHIFT_IP" \
+        "sudo /etc/greenboot/check/required.d/40_microshift_running_check.sh | \
+        while IFS= read -r line; do printf '%s %s\\n' \"\$(date +'%H:%M:%S.%N')\" \"\$line\"; done" &>"${test_output}/0002-readiness-check.log"
+}
+
+microshift_setup() {
+    local test_output="${1}"
+    log "Setting up and starting MicroShift"
+    ssh "$USHIFT_USER@$USHIFT_IP" 'cat << EOF | sudo tee /etc/microshift/config.yaml
+---
+apiServer:
+  subjectAltNames:
+  - '"$USHIFT_IP"'
+EOF' &>"${test_output}/0001-setup.log"
+    ssh "$USHIFT_USER@$USHIFT_IP" "sudo systemctl enable --now microshift" &>>"${test_output}/0001-setup.log"
```

Comment:
That was my initial approach, but then piping whole function output to file leads to messages like "setting up microshift" to be in that file instead of printed out to console. Unless there's some magic with streams (like "separate runner's std{err,out} from test's std{err,out})?

Slightly related: I consider this script as an intermediate step - move tests here, see what we need, what we want.
I hope that we'll be able to move to some other solution. I spent some time over the weekend [PoCing python script](https://github.com/pmtk/microshift/commit/d25d9f6ff4b81c9dcefe05471dc8b0ab8c29fb4a) and I see needs beside what we have here:
- junit xml is very nice addition
- tagging/labeling tests (e.g. to group "non-destructive" like smoke tests we have now together without tearing down and setting up microshift in between; as opposed to greenboot test that tears down microshift multiple times)

Reply:
OK, let's keep things as you have them for now, then.